supervisor/iptablesctrl/uidpam/update_portsets.go

```diff
@@ -0,0 +1,260 @@
+func (p *UIDCacheInstance) UpdatePortSet(username string, port string) (err error) {
+
+	puPortSetName, err := p.GetFromUIDToPortSetCache(username)
+	if err != nil {
+		return fmt.Errorf("Unable to get portset from uid")
+	}
+
+	ips := ipset.IPSet{
+		Name: puPortSetName,
+	}
+
+	if _, err := strconv.Atoi(port); err != nil {
+		return fmt.Errorf("Not a valid Port")
+	}
+
+	if adderr := ips.Add(port, portEntryTimeout); adderr != nil {
+		return fmt.Errorf("Unable to add port to set")
+	}
+	return nil
+}
+
+// InitPortSetTask This go routine periodically scans (1s)
+// /proc/net/tcp file for listening ports and programs
+// the portsets. This worker thread is setup during datapath
+// initilisation.
+func InitPortSetTask(p *UIDCacheInstance) {
+	go startPortSetTask(p)
+}
+
+func startPortSetTask(p *UIDCacheInstance) {
+	t := time.NewTicker(portSetUpdateIntervalMilliseconds * time.Millisecond)
+	for range t.C {
+		// Update PortSet periodically.
+		p.updateIPPortSets()
+	}
+}
+
+func (p *UIDCacheInstance) updateIPPortSets() {
+	file, err := os.Open(procNetTCPFile)
+	if err != nil {
+		zap.L().Warn("Failed To open /proc/net/tcp file", zap.Error(err))
+		return
+	}
+	defer func() {
+		err := file.Close()
+		if err != nil {
+			zap.L().Warn("Failed To close /proc/net/tcp file", zap.Error(err))
+		}
+	}()
+
+	localCache := cache.NewCache("localCache")
+	scanner := bufio.NewScanner(file)
+	lineCnt := 0
+	for scanner.Scan() {
+		line := strings.Fields(scanner.Text())
+		lineCnt++
+		portList := make([]int64, 0)
+		if (lineCnt == procHeaderLineNum) || (line[sockStateOffset] != sockListeningState) {
+			continue
+		}
+		uid := line[uidFieldOffset]
+		portStr := strings.Split(line[ipPortOffset], ":")[portOffset]
+
+		portNum, err := strconv.ParseInt(portStr, hexFormat, integerSize)
+		if err != nil {
+			zap.L().Warn("Failed to convert port to Int", zap.Error(err))
+			return
+		}
+
+		if v, err := localCache.Get(uid); err == nil {
+			portList = append(v.([]int64), int64(portNum))
+		} else {
+			portList = append(portList, int64(portNum))
+		}
+
+		localCache.AddOrUpdate(uid, portList)
+
+	}
+
+	if err := scanner.Err(); err != nil {
+		zap.L().Warn("Error while parsing /proc/net/tcp ", zap.Error(err))
+		return
+
+	}
+
+	p.UIDToPorts = localCache
+
+	// program the iptable.
+	activePUs := p.UIDSet
+	for k := range activePUs {
+		uidPorts, err := p.UIDToPorts.Get(k) //getUIDPortMappings(k)
+		if err != nil {
+			continue
+		}
+		puPortSetName, err := p.UIDToPortSet.Get(k) //getUIDPortSetMappings(k)
+
+		if err != nil {
+			// Not Fatal. A normal uidlogin pu will not have a port
+			continue
+		}
+
+		// ipset
+		ips := ipset.IPSet{
+			Name: puPortSetName.(string),
+		}
+		// rearm the portset with new port List
+		for _, port := range uidPorts.([]int64) {
+			//Add an entry for 60 seconds we will rediscover ports every 60 sec
+			portNum := strconv.Itoa(int(port))
+			if adderr := ips.Add(portNum, portEntryTimeout); adderr != nil {
```

Comment:
Whats this timeout ? We should be adding/removing.

Reply:
Removing the port from portset would require iterating over list of ports for each active uid and comparing with its previous list to determine port that needs to be removed. We know a port is no longer active when the entry no longer appears in /proc/net/tcp. 

Since we update the portset every second, we will always have a right portsets installed in the ipset.